rust/worker/src/execution/operators/metadata_filtering.rs

```diff
@@ -0,0 +1,1348 @@
+                Some(meta) => {
+                    for (meta_key, meta_val) in meta {
+                        map_pointer.insert(meta_key, meta_val);
+                    }
+                }
+                None => (),
+            };
+        }
+        let clo = |metadata_key: &str,
+                   metadata_value: &crate::blockstore::key::KeyWrapper,
+                   metadata_type: crate::types::MetadataType,
+                   comparator: WhereClauseComparator| {
+            match metadata_type {
+                crate::types::MetadataType::StringType => match comparator {
+                    WhereClauseComparator::Equal => {
+                        let mut result = RoaringBitmap::new();
+                        // Construct a bitmap consisting of all offset ids
+                        // that have this key equal to this value.
+                        for (offset_id, meta_map) in &ids_to_metadata {
+                            if let Some(val) = meta_map.get(&metadata_key.to_string()) {
+                                match *val {
+                                    MetadataValue::Str(string_value) => {
+                                        if let KeyWrapper::String(where_value) = metadata_value {
+                                            if *string_value == *where_value {
+                                                result.insert(*offset_id);
+                                            }
+                                        }
+                                    }
+                                    _ => (),
+                                }
+                            }
+                        }
+                        return result;
+                    }
+                    WhereClauseComparator::NotEqual => {
+                        todo!();
+                    }
+                    // We don't allow these comparators for strings.
+                    WhereClauseComparator::LessThan => {
+                        unimplemented!();
+                    }
+                    WhereClauseComparator::LessThanOrEqual => {
+                        unimplemented!();
+                    }
+                    WhereClauseComparator::GreaterThan => {
+                        unimplemented!();
+                    }
+                    WhereClauseComparator::GreaterThanOrEqual => {
+                        unimplemented!();
+                    }
+                },
+                crate::types::MetadataType::IntType => match comparator {
+                    WhereClauseComparator::Equal => {
+                        let mut result = RoaringBitmap::new();
+                        // Construct a bitmap consisting of all offset ids
+                        // that have this key equal to this value.
+                        for (offset_id, meta_map) in &ids_to_metadata {
+                            if let Some(val) = meta_map.get(&metadata_key.to_string()) {
+                                match *val {
```

Comment:
why are we taking a reference of a to_string()? to_string() will clone the key and then we take a reference?